src/common/setting_internal.h

```diff
@@ -0,0 +1,216 @@
+};
+
+template<typename T>
+constexpr auto GetPtr(std::optional<T>(*ptr)(const common::SettingsValue&)) { return ptr; }
+
+template<SettingOptions options>
+struct SettingTraits<bool, options> : SettingTraitsBase<bool, GetPtr(SettingToBool)> {};
+
+template<typename T, SettingOptions options>
+requires std::is_integral_v<T>
+struct SettingTraits<T, options> : SettingTraitsBase<T, GetPtr(SettingToInt)> {};
+
+template<SettingOptions options>
+struct SettingTraits<std::string, options> : SettingTraitsBase<std::string, GetPtr(SettingToString)> {};
+
+template<SettingOptions options>
+struct SettingTraits<fs::path, options> : SettingTraitsBase<fs::path, &SettingToPath> {};
+
+template<typename T, SettingOptions options>
+struct SettingTraits<std::optional<T>, options> : SettingTraits<T, options>
+{
+    using setting_t = std::optional<T>;
+    using wrapped_t = SettingTraits<T, options>;
+    static constexpr bool is_list{wrapped_t::is_list};
+
+    static bool Get(const common::SettingsValue& setting, setting_t& out)
+    {
+        bool unset{!out};
+        if (unset) out.emplace();
+        bool got{wrapped_t::Get(setting, *out)};
+        if (unset && !got) out.reset();
+        return got;
+    }
+};
+
+template<typename T, SettingOptions options>
+struct SettingTraits<std::vector<T>, options> : SettingTraits<T, options>
+{
+    using setting_t = std::vector<T>;
+    using wrapped_t = SettingTraits<T, options>;
+    static constexpr bool is_list{true};
+
+    static bool Get(const common::SettingsValue& setting, setting_t& out)
+    {
+        T elem;
+        if (wrapped_t::Get(setting, elem)) {
+            out.emplace_back(std::move(elem));
+            return true;
+        }
+        return false;
+    }
+};
+
+template<SettingOptions options>
+struct SettingTraits<SettingsValue, options> : SettingTraitsBase<SettingsValue>
+{
+    static bool Get(const common::SettingsValue& setting, setting_t& out) { out = setting; return true; }
+};
+
+inline std::string_view SettingName(std::string_view summary)
+{
+    return summary.substr(0, summary.find_first_of('='));
+}
+
+template<SettingOptions options>
+constexpr int SettingFlags()
+{
+    int flags = 0;
+    if (options.debug_only)        flags |= ArgsManager::DEBUG_ONLY;
+    if (options.network_only)      flags |= ArgsManager::NETWORK_ONLY;
+    if (options.sensitive)         flags |= ArgsManager::SENSITIVE;
+    if (options.disallow_negation) flags |= ArgsManager::DISALLOW_NEGATION;
+    if (options.disallow_elision)  flags |= ArgsManager::DISALLOW_ELISION;
+    return flags;
+}
+
+template<auto _value>
+struct Constant {
+    static constexpr auto value = _value;
+    using type = decltype(value);
+    constexpr auto operator()() { return value; }
+};
+
+template<typename T, SettingOptions options, auto help>
+void SettingRegister(auto& manager, auto summary, auto category, auto help_fn, auto default_fn, auto get_fn, auto&&... register_options)
```

Comment:
Document types and remove unused `T`?
```suggestion
template<SettingOptions options, auto help>
void SettingRegister(ArgsManager& manager, auto summary, OptionsCategory category, auto help_fn, auto default_fn, auto get_fn, auto&&... register_options)
```